### PR TITLE
Connect other peer via multiple ports (review suggestion applied)

### PR DESCRIPTION
**Pull request by mo-esmp**

### CODE REVIEW

File: Program.cs [hunk 1]
```diff
@@ -104,11 +101,12 @@ static void Main(string[] args)
...
             finally
             {
-                socket.Close();
+                foreach (var socket in Sockets)
+                    socket?.Close();
             }
         }
 
-        class CommandLineArguments
+        private class CommandLineArguments
```
Comment (knocte) [inline Program.cs]: this line is not changing behaviour, so let's not add private keyword here, we can do that in a refactoring PR

File: Program.cs [hunk 2]
```diff
@@ -1,15 +1,22 @@
...
+using System.Net.Sockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
 using UdtSharp;
 
 namespace p2pcopy
 {
-    class Program
+    internal class Program
```
Comment (knocte) [inline Program.cs]: this line is not changing behaviour, so let's not add private keyword here, we can do that in a refactoring PR

File: Program.cs [hunk 3]
```diff
@@ -471,65 +480,127 @@ static P2pEndPoint GetExternalEndPoint(Socket socket)
...
+                return;
+            }
+
+            Console.WriteLine();
+            Console.WriteLine("Tell your peer your IP is: {0}", p2pEndPoint.External.Address);
+            Console.WriteLine("Initializing sockets...");
+
+            var ports = new StringBuilder();
+
+            for (var i = 0; i < 10; i++)
```
Comment (knocte) [inline Program.cs]: can you store the number 10 in a constant so that it's easily findable and changeable later please

File: StunClient.cs [hunk 4]
```diff
@@ -825,7 +826,7 @@ internal class StunClient
         /// <param name="socket">UDP socket to use.</param>
         /// <returns>Returns UDP netwrok info.</returns>
         /// <exception cref="Exception">Throws exception if unexpected error happens.</exception>
-        internal static StunResult Query(string host, int port, Socket socket)
+        internal static async Task<StunResult> QueryAsync(string host, int port, Socket socket)
```
Comment (knocte) [inline StunClient.cs]: can we leave changes to the StunClient away from this PR? we can leave them for another PR

File: Program.cs (hunk 1 above)
Comment (mo-esmp) [reply Program.cs]: Ah, I have several extensions like Resharper, CodeMaid ... that fix and clean up codes. I will fix this via new commit

File: Program.cs (hunk 2 above)
Comment (mo-esmp) [reply Program.cs]: I will fix this via new commit

File: Program.cs (hunk 3 above)
Comment (mo-esmp) [reply Program.cs]: Ok sure

File: StunClient.cs (hunk 4 above)
Comment (mo-esmp) [reply StunClient.cs]: I have to use the async overload of the `Query` method, otherwise I have to change the implementation of `ConnectAsync` to

```
static Task<UdtSocket> ConnectAsync(Socket socket, string remoteAddress, int remotePort, CancellationToken ct)
{
    try
    {
        GetExternalEndPointAsync(socket);

        var client = new UdtSocket(socket.AddressFamily, socket.SocketType);
        client.Bind(socket);

        Console.WriteLine("Trying to connect to {0}:{1}.  ", remoteAddress, remotePort);

        client.Connect(new IPEndPoint(IPAddress.Parse(remoteAddress), remotePort));

        Console.WriteLine("Connected successfully to {0}:{1}", remoteAddress, remotePort);

        if (ct.IsCancellationRequested)
        {
            socket.Close();
            ct.ThrowIfCancellationRequested();
        }

        return Task.FromResult(client);
    }
```

And all the tasks

```var taskList = remotePorts .Select((port, index) => ConnectAsync(Sockets[index], remoteAddress, port, cts.Token)).ToList();```

Will be executed synchronously.

Comment (knocte) [reply StunClient.cs]: I understand that asynchronicity is better, but in this case it doesn't help because p2pcopy is a command-line tool anyway (so there's no need to send workload to the background while the UI runs) so there's not much difference here. Especially considering that the part that this PR needs to change is the connection between peers, not the connection to Stun servers (which worked fine, even if synchronously). We can make the connection to STUN asynchronous later in another PR.

**Comment by mo-esmp** [comment]
I added a new commit and fixed access modifiers and the constant number.

File: StunClient.cs (hunk 4 above)
Comment (mo-esmp) [reply StunClient.cs]: The purpose of this PR was to parallelize the connection to the client through different ports. By the way, I can remove all async codes and just try each port, one after another synchronously if you think there's not much difference between sync and async here.

Comment (knocte) [reply StunClient.cs]: > The purpose of this PR was to parallelize the connection to the client through different ports.

Parallelize the connection to the peer, not to the STUN server.

> By the way, I can remove all async codes and just try each port, one after another synchronously if you think there's not much difference between sync and async here.

No, leave the connection to the peer asynchronous as you have already refactored it like that. But don't touch Stun-related code, it's all am asking.

Comment (mo-esmp) [reply StunClient.cs]: `UdtSocket` itself doesn't have any async operation and `STUN` server makes a transaction with the `TcpSocket` so I have to make this operation async to run tasks in the background.  As I posted the other version of `ConnectAsync` that returns `Task.FromResult(client);`, it will be executed synchronously and it doesn't help much. Another way is changing `ConnectAsync` to sync the version and running it with `Task.Run` in the background.

Comment (knocte) [reply StunClient.cs]: I don't understand why you need to run any STUN-related stuff in the background, this was never asked for.

Comment (knocte) [reply StunClient.cs]: (The 10 threads/tasks thing shouldn't start until after the connection with STUN servers has ended.)

Comment (mo-esmp) [reply StunClient.cs]: Could you please review my last two commits?

Comment (knocte) [reply StunClient.cs]: looking good! I think we're only missing a couple of nitpicks more

File: Program.cs [hunk 5]
```diff
@@ -177,7 +184,7 @@ class P2pEndPoint
             internal IPEndPoint Internal;
         }
 
-        static P2pEndPoint GetExternalEndPoint(Socket socket)
+        static P2pEndPoint GetExternalEndPoint(Socket socket, bool printMessages = true)
```
Comment (knocte) [inline Program.cs]: this change above seems unrelated to the task at hand

File: Program.cs [hunk 6]
```diff
@@ -52,33 +52,31 @@ static void Main(string[] args)
...
+                if (string.IsNullOrEmpty(remoteIp))
                 {
-                    Console.WriteLine("Invalid ip:port entered");
+                    Console.WriteLine("Invalid ip entered");
                     return;
                 }
 
-                // try again to connect to external to "reopen" port
-                GetExternalEndPoint(socket);
+                Console.Write("Enter the ports of your peer (comma separate): ");
```
Comment (knocte) [inline Program.cs]: why change the UX of the program? you made the user have to do an extra copy+paste operation; but I think it's not needed: you should only ask about the IP and ports once, but instead of the format being ip:port, the format will be ip:port,port,port,..,port

File: Program.cs [hunk 7]
```diff
@@ -1,15 +1,23 @@
...
+using System.Threading;
+using System.Threading.Tasks;
 using UdtSharp;
 
 namespace p2pcopy
 {
     class Program
     {
-        static void Main(string[] args)
+        private static int _defaultTaskCount = 10;
```
Comment (knocte) [inline Program.cs]: nit: I believe the existing coding style of p2pcopy doesn't use underscores for fields

File: Program.cs [hunk 8]
```diff
@@ -1,15 +1,23 @@
...
+        private static readonly List<Socket> Sockets = new List<Socket>();
```
Comment (knocte) [inline Program.cs]: nit: I believe the current coding style of p2pcopy uses camelCase (not PascalCase) for fields

File: Program.cs (hunk 7 above)
Comment (mo-esmp) [reply Program.cs]: Couldn't find class-level private field in the codebase, however, I'll remove the underscore.

File: Program.cs (hunk 8 above)
Comment (mo-esmp) [reply Program.cs]: Well, I follow the Resharper naming convention, for `static readonly` fields it suggests PC, I'll change to cC.

File: Program.cs (hunk 6 above)
Comment (mo-esmp) [reply Program.cs]: OK

File: Program.cs (hunk 5 above)
Comment (mo-esmp) [reply Program.cs]: Ok, I rollback this change.

File: Program.cs [hunk 9]
```diff
@@ -470,66 +472,123 @@ static P2pEndPoint GetExternalEndPoint(Socket socket)
...
+            
             Console.WriteLine("Could not find a working STUN server");
+
             return null;
         }
 
-
         static int SleepTime(DateTime now)
         {
-            List<int> seconds = new List<int>() {10, 20, 30, 40, 50, 60};
```
Comment (knocte) [inline Program.cs:481]: any reason why moving this line of code? seems it's not used anywhere else so it doesn't need to move

File: Program.cs [hunk 10]
```diff
@@ -470,66 +472,123 @@ static P2pEndPoint GetExternalEndPoint(Socket socket)
...
 
-                    if (client != null)
-                        client.Close();
+                Console.WriteLine("[{0}] - Waiting {1} sec to sync with other peer", now.ToLongTimeString(), sleepTimeToSync);
 
-                    client = new UdtSocket(socket.AddressFamily, socket.SocketType);
+                await Task.Delay(TimeSpan.FromSeconds(sleepTimeToSync), cts.Token);
+
+                Console.WriteLine("Please wait... ");
+                udtSocket = await (await Task.WhenAny(taskList));
```
Comment (knocte) [inline Program.cs]: two consecutive awaits? this looks weird

File: Program.cs (hunk 9 above)
Comment (mo-esmp) [reply Program.cs:481]: I guess I failed the test.

File: Program.cs [hunk 11]
```diff
@@ -470,66 +472,123 @@ static P2pEndPoint GetExternalEndPoint(Socket socket)
...
-            int retry = 0;
+            if (!sockets.Any())
+                return null;
+
+            Console.WriteLine("Trying to connect to peer on different ports");
 
-            UdtSocket client = null;
+            UdtSocket udtSocket;
+            CancellationTokenSource cts;
+            var retry = 1;
```
Comment (knocte) [inline Program.cs]: why make retry start at 1 instead of 0?

File: Program.cs (hunk 10 above)
Comment (mo-esmp) [reply Program.cs]: First await return the task and the second await gets result from the task.

File: Program.cs (hunk 11 above)
Comment (mo-esmp) [reply Program.cs:497]: Copy/Paste bug.

Comment (mo-esmp) [reply Program.cs]: After the `WriteLine` retry will be increased, there no problem with the code, however, it could be written in this way
```csharp
var retry = 0;
Console.WriteLine($"Try {++retry}:");
```

File: Program.cs (hunk 9 above)
Comment (knocte) [reply Program.cs:481]: hehe TBH I don't know yet if you passed the test because I didn't run your code yet (but I will soon); I'm just being very picky with the changes to get the PR ready to be proposed to the original repo

File: Program.cs (hunk 11 above)
Comment (knocte) [reply Program.cs]: you look to be talking about a refactoring?, which should be left for another PR; this PR is only to implement the 10 thread/task behaviour change

File: Program.cs (hunk 9 above)
Comment (mo-esmp) [reply Program.cs:481]: I guess there will not be any difference between the result of code reviews and running the application, don't expect much from hitting F5.

File: Program.cs (hunk 11 above)
Comment (mo-esmp) [reply Program.cs]: I admit I failed the test, feel free to close the PR.

Comment (knocte) [reply Program.cs]: If you keep being this negative it is when you will fail the test :) You were doing good so far ;) This is just a review, which is part of the test, hehe.

Comment (mo-esmp) [reply Program.cs]: I'm not upset with the failure. Most of the time failure is part of success. I've learned a lot from this review, thanks.

File: Program.cs (hunk 9 above)
Comment (knocte) [reply Program.cs:481]: I didn't say that I expect a different behaviour from before the reviews to after the reviews, I mean that I even haven't run your code yet

File: Program.cs [hunk 12]
```diff
@@ -470,66 +471,125 @@ static P2pEndPoint GetExternalEndPoint(Socket socket)
...
+                Console.WriteLine("[{0}] - Waiting {1} sec to sync with other peer", now.ToLongTimeString(), sleepTimeToSync);
 
-                    if (client != null)
-                        client.Close();
+                await Task.Delay(TimeSpan.FromSeconds(sleepTimeToSync), cts.Token);
 
-                    client = new UdtSocket(socket.AddressFamily, socket.SocketType);
+                Console.WriteLine("Please wait... ");
+                udtSocket = await (await Task.WhenAny(taskList));
+            } while (udtSocket == null);
```
Comment (knocte) [inline Program.cs]: how could udtSocket ever be null?

Comment (mo-esmp) [reply Program.cs]: In the ConnectAsync method, if an exception is thrown, null is returned.

Comment (knocte) [reply Program.cs]: but then shouldn't the do..while be instead in the task that calls ConnectAsync?

Comment (knocte) [reply Program.cs]: otherwise, it could happen that the fastest task to execute is the one that raises the exception and receives null, which then makes the WhenAny select the faulty one, which doesn't seem to be the best outcome

Comment (mo-esmp) [reply Program.cs]: If the socket fails to establish a connection, it will take several seconds for an exception to be thrown (after a timeout) and the null to be returned. If the first null is returned, the connecting process is performed again. If the first socket succeeds in connecting, the socket value will be returned and the rest of the tasks will be canceled.

Comment (knocte) [reply Program.cs]: >  it will take several seconds for an exception to be thrown (after a timeout) and the null to be returned

That's a very big assumption

> If the first null is returned, the connecting process is performed again.

Launch the 10 threads/tasks again, when the other 9 are already running and well? that seems very efficient.

I'm thinking that if you don't capture the exception to return null (so instead let it throw), then the worst that will happen is that Task.WhenAny() will not take in account that task (because it didn't succeed). This could be more efficient.

Comment (mo-esmp) [reply Program.cs]: If I don't capture the exception, the console app will crash. The task will not necessarily run on the new thread, and the main thread of the app may run the tasks.
image

`WhenAny` can be replaced with `WhenAll` but if you run the app you will get noticed that if a socket fails to connect it takes several seconds for an exception to be thrown, while another socket connected successfully and has to wait for the result of other sockets.

Comment (mo-esmp) [reply Program.cs:518]: Any ideas?

Comment (knocte) [reply Program.cs]: WhenAll would defeat the purpose of this change.

Let me think about this today and test a couple of things.

Comment (knocte) [reply Program.cs]: Sorry for the delay Moshen! How about if we use this? https://stackoverflow.com/a/49313971/544947

Comment (mo-esmp) [reply Program.cs]: I made the changes، please check out [1acbc3f](https://github.com/nblockchain/p2pcopy/pull/2/commits/1acbc3fcfb1337e386aa612cee910f4113113470).

Comment (knocte) [reply Program.cs]: looks great, I'll test it today or tomorrow

**Comment by knocte** [comment]
Finally I found time last nite to run/test this PR.

I have the following observations:
- It seems the STUN client code is being hit 10 times, when it should be still 1, because this got printed:
```
Contacting STUN servers to obtain your IP
Your firewall is FullCone
Contacting STUN servers to obtain your IP
Your firewall is FullCone
Contacting STUN servers to obtain your IP
Your firewall is FullCone
Contacting STUN servers to obtain your IP
Your firewall is FullCone
Contacting STUN servers to obtain your IP
Your firewall is FullCone
Contacting STUN servers to obtain your IP
Your firewall is FullCone
Contacting STUN servers to obtain your IP
Your firewall is FullCone
Contacting STUN servers to obtain your IP
Your firewall is FullCone
Contacting STUN servers to obtain your IP
Your firewall is FullCone
```
- With the previous code (before your PR), if the connection was not successful at first, it would keep retrying. However, after your PR, if the first attempt fails (the 10 threads), it fails.
- Somehow the connections don't seem to be really in parallel because when I ran it, this got printed first (and it hang there for a while until the next lines got printed):
```
Start connecting via sockets, please wait... 
Try 0:
```

**Comment by mo-esmp** [comment]
- In the `InitializeSockets` method 10 sockets will be initialized. for each socket, I call `GetExternalEndPoint` to obtain IP and port (`StunClient.Query(host, port, socket);`) and `Contacting STUN servers to obtain your IP` is printed. I don't know how to get 10 ports with one round trip
- Retrying logic has broken after last the commit, I'll fix it
- As I said earlier if a socket fails to establish a connection, it will take several seconds for an exception to be thrown (after a timeout) and that is why the connection appears to be hanged. If all 10 sockets fail to establish a connection, they all throw exception almost at the same time (due to socket default timeout)

**Comment by knocte** [comment]
> In the InitializeSockets method, 10 sockets will be initialized. for each socket, I call GetExternalEndPoint to obtain IP and port (StunClient.Query(host, port, socket)

Don't call GetExternalEndPoint after launching the 10threads/tasks, do it before. GetExternalEndPoint should only get called once.

**Comment by knocte** [comment]
(Because I assume GetExternalEndPoint just gets the public IP via STUN, and it doesn't do anything about ports.)

**Comment by mo-esmp** [comment]
> (Because I assume GetExternalEndPoint just gets the public IP via STUN, and it doesn't do anything about ports.)

Yes you right, my implementation was based on a false assumption.
